File: Phyton-Service/shared/rate_limiter.py
"""
🚦 RATE LIMITER
===============
Request rate limiting for API endpoints

Features:
- IP-based rate limiting
- Configurable limits per endpoint
- Automatic cleanup of old requests
- Thread-safe implementation

WHITE-HAT COMPLIANCE: DoS attack prevention (not malicious throttling)
"""

from functools import wraps
from flask import request, jsonify
import time
from collections import defaultdict
from threading import Lock
from typing import Callable


class RateLimiter:
    """
    Rate limiter with sliding window algorithm

    Example:
        limiter = RateLimiter(requests_per_minute=60)
        if limiter.is_allowed("192.168.1.1"):
            # Process request
        else:
            # Return 429 Too Many Requests
    """

    def __init__(self, requests_per_minute: int = 60):
        self.requests_per_minute = requests_per_minute
        self.requests = defaultdict(list)
        self.lock = Lock()

    def is_allowed(self, identifier: str) -> bool:
        """
        Check if request is allowed for given identifier

        Args:
            identifier: Usually IP address or user ID

        Returns:
            True if request is allowed, False if rate limit exceeded
        """
        with self.lock:
            now = time.time()
            minute_ago = now - 60

            # Clean old requests (older than 1 minute)
            self.requests[identifier] = [
                req_time for req_time in self.requests[identifier]
                if req_time > minute_ago
            ]

            # Check limit
            if len(self.requests[identifier]) >= self.requests_per_minute:
                return False

            # Add new request
            self.requests[identifier].append(now)
            return True

    def get_remaining(self, identifier: str) -> int:
        """Get remaining requests for identifier"""
        with self.lock:
            now = time.time()
            minute_ago = now - 60

            recent_requests = [
                req_time for req_time in self.requests.get(identifier, [])
                if req_time > minute_ago
            ]

            return max(0, self.requests_per_minute - len(recent_requests))


def rate_limit(requests_per_minute: int = 60):
    """
    Decorator for rate limiting Flask routes

    Args:
        requests_per_minute: Maximum requests allowed per minute per IP

    Example:
        @app.route('/api/signals')
        @rate_limit(requests_per_minute=100)
        def get_signals():
            return jsonify({'data': []})
    """
    limiter = RateLimiter(requests_per_minute)

    def decorator(f: Callable) -> Callable:
        @wraps(f)
        def decorated_function(*args, **kwargs):
            # IP tabanlı rate limiting
            identifier = request.remote_addr or 'unknown'

            if not limiter.is_allowed(identifier):
                remaining_time = 60  # seconds

                return jsonify({
                    'success': False,
                    'error': 'Rate limit exceeded',
                    'message': f'Maximum {requests_per_minute} requests per minute allowed',
                    'retry_after': remaining_time
                }), 429

            # Add rate limit info to response headers
            response = f(*args, **kwargs)

            if hasattr(response, 'headers'):
                remaining = limiter.get_remaining(identifier)
                response.headers['X-RateLimit-Limit'] = str(requests_per_minute)
                response.headers['X-RateLimit-Remaining'] = str(remaining)
                response.headers['X-RateLimit-Reset'] = str(int(time.time()) + 60)

            return response

        return decorated_function
    return decorator


# Example usage
if __name__ == "__main__":
    # Test rate limiter
    limiter = RateLimiter(requests_per_minute=5)

    test_ip = "192.168.1.1"

    print("Testing rate limiter (5 requests/minute):")
    for i in range(7):
        if limiter.is_allowed(test_ip):
            remaining = limiter.get_remaining(test_ip)
            print(f"✅ Request {i+1} allowed ({remaining} remaining)")
        else:
            print(f"❌ Request {i+1} blocked (rate limit exceeded)")

File: src/lib/security/rate-limiter.ts
/**
 * 🛡️ ADVANCED RATE LIMITER
 * Multi-strategy rate limiting for API protection
 *
 * Features:
 * - IP-based rate limiting
 * - User-based rate limiting
 * - Endpoint-based rate limiting
 * - Sliding window algorithm
 * - Distributed support (Redis)
 * - Automatic cleanup
 * - White-hat compliance: All violations logged
 */

// ============================================================================
// TYPES
// ============================================================================

export interface RateLimitConfig {
  /**
   * Maximum requests per window
   */
  maxRequests: number;

  /**
   * Time window in milliseconds
   */
  windowMs: number;

  /**
   * Rate limit key prefix (for namespacing)
   */
  keyPrefix?: string;

  /**
   * Skip rate limiting for certain conditions
   */
  skip?: (identifier: string) => boolean;
}

export interface RateLimitResult {
  allowed: boolean;
  remaining: number;
  resetAt: number;
  retryAfter?: number;
}

interface RateLimitEntry {
  count: number;
  resetAt: number;
  requests: number[]; // Timestamps for sliding window
}

// ============================================================================
// RATE LIMITER CLASS
// ============================================================================

export class RateLimiter {
  private store: Map<string, RateLimitEntry> = new Map();
  private config: RateLimitConfig;
  private cleanupInterval: NodeJS.Timeout | null = null;

  constructor(config: RateLimitConfig) {
    this.config = {
      keyPrefix: 'ratelimit',
      ...config,
    };

    // Start automatic cleanup (every 5 minutes)
    this.cleanupInterval = setInterval(() => {
      this.cleanup();
    }, 300000);

    console.log(
      `[RateLimiter] Initialized: ${config.maxRequests} req/${config.windowMs}ms`
    );
  }

  /**
   * Check rate limit for identifier
   */
  check(identifier: string): RateLimitResult {
    // Skip if configured
    if (this.config.skip && this.config.skip(identifier)) {
      return {
        allowed: true,
        remaining: this.config.maxRequests,
        resetAt: Date.now() + this.config.windowMs,
      };
    }

    const key = `${this.config.keyPrefix}:${identifier}`;
    const now = Date.now();

    // Get or create entry
    let entry = this.store.get(key);

    if (!entry) {
      entry = {
        count: 0,
        resetAt: now + this.config.windowMs,
        requests: [],
      };
      this.store.set(key, entry);
    }

    // Clean old requests (sliding window)
    entry.requests = entry.requests.filter(
      (timestamp) => now - timestamp < this.config.windowMs
    );

    // Check if limit exceeded
    if (entry.requests.length >= this.config.maxRequests) {
      const oldestRequest = entry.requests[0];
      const retryAfter = Math.ceil((oldestRequest + this.config.windowMs - now) / 1000);

      // Log violation
      console.warn(
        `[RateLimiter] Limit exceeded: ${identifier} (${entry.requests.length}/${this.config.maxRequests})`
      );

      return {
        allowed: false,
        remaining: 0,
        resetAt: oldestRequest + this.config.windowMs,
        retryAfter,
      };
    }

    // Add current request
    entry.requests.push(now);
    entry.count++;

    const remaining = this.config.maxRequests - entry.requests.length;

    return {
      allowed: true,
      remaining,
      resetAt: entry.requests[0] + this.config.windowMs,
    };
  }

  /**
   * Increment counter for identifier
   */
  increment(identifier: string): RateLimitResult {
    return this.check(identifier);
  }

  /**
   * Reset rate limit for identifier
   */
  reset(identifier: string): void {
    const key = `${this.config.keyPrefix}:${identifier}`;
    this.store.delete(key);
    console.log(`[RateLimiter] Reset: ${identifier}`);
  }

  /**
   * Get current count for identifier
   */
  getCount(identifier: string): number {
    const key = `${this.config.keyPrefix}:${identifier}`;
    const entry = this.store.get(key);
    return entry ? entry.requests.length : 0;
  }

  /**
   * Cleanup expired entries
   */
  cleanup(): void {
    const now = Date.now();
    let removedCount = 0;

    for (const [key, entry] of this.store.entries()) {
      // Remove if all requests are older than window
      entry.requests = entry.requests.filter(
        (timestamp) => now - timestamp < this.config.windowMs
      );

      if (entry.requests.length === 0) {
        this.store.delete(key);
        removedCount++;
      }
    }

    if (removedCount > 0) {
      console.log(`[RateLimiter] Cleaned up ${removedCount} expired entries`);
    }
  }

  /**
   * Get statistics
   */
  getStats(): {
    totalEntries: number;
    totalRequests: number;
    config: RateLimitConfig;
  } {
    let totalRequests = 0;

    for (const entry of this.store.values()) {
      totalRequests += entry.requests.length;
    }

    return {
      totalEntries: this.store.size,
      totalRequests,
      config: this.config,
    };
  }

  /**
   * Shutdown rate limiter
   */
  shutdown(): void {
    if (this.cleanupInterval) {
      clearInterval(this.cleanupInterval);
      this.cleanupInterval = null;
    }

    this.store.clear();
    console.log('[RateLimiter] Shutdown complete');
  }
}

// ============================================================================
// PRE-CONFIGURED RATE LIMITERS
// ============================================================================

/**
 * Global API rate limiter (60 req/min per IP)
 * Development: Unlimited for localhost (::1, 127.0.0.1, unknown)
 * Production: 600 req/min for normal traffic
 */
export const globalRateLimiter = new RateLimiter({
  maxRequests: parseInt(process.env.RATE_LIMIT_MAX_REQUESTS || '600', 10), // Increased to 600 for development
  windowMs: parseInt(process.env.RATE_LIMIT_WINDOW_MS || '60000', 10),
  keyPrefix: 'global',
  skip: (identifier: string) => {
    // Skip rate limiting for localhost during development
    const isLocalhost = identifier === '::1' || identifier === '127.0.0.1' || identifier === 'unknown';
    const isDevelopment = process.env.NODE_ENV === 'development';
    return isLocalhost && isDevelopment;
  },
});

/**
 * Strict rate limiter for sensitive endpoints (100 req/min per IP)
 * Development: Unlimited for localhost
 */
export const strictRateLimiter = new RateLimiter({
  maxRequests: 100, // Increased for development
  windowMs: 60000, // 1 minute
  keyPrefix: 'strict',
  skip: (identifier: string) => {
    const isLocalhost = identifier === '::1' || identifier === '127.0.0.1' || identifier === 'unknown';
    const isDevelopment = process.env.NODE_ENV === 'development';
    return isLocalhost && isDevelopment;
  },
});

/**
 * Auth rate limiter for login/register (50 req/min per IP)
 * Development: Unlimited for localhost
 */
export const authRateLimiter = new RateLimiter({
  maxRequests: 50, // Increased for development
  windowMs: 60000, // 1 minute
  keyPrefix: 'auth',
  skip: (identifier: string) => {
    const isLocalhost = identifier === '::1' || identifier === '127.0.0.1' || identifier === 'unknown';
    const isDevelopment = process.env.NODE_ENV === 'development';
    return isLocalhost && isDevelopment;
  },
});

/**
 * Scanner enqueue rate limiter (100 req/min per user)
 * Development: Unlimited for localhost
 */
export const scannerRateLimiter = new RateLimiter({
  maxRequests: 100, // Increased for development
  windowMs: 60000, // 1 minute
  keyPrefix: 'scanner',
  skip: (identifier: string) => {
    const isLocalhost = identifier === '::1' || identifier === '127.0.0.1' || identifier === 'unknown';
    const isDevelopment = process.env.NODE_ENV === 'development';
    return isLocalhost && isDevelopment;
  },
});

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

/**
 * Get client IP address from request
 */
export function getClientIP(request: Request): string {
  const headers = request.headers;

  // Try common proxy headers
  const forwardedFor = headers.get('x-forwarded-for');
  if (forwardedFor) {
    return forwardedFor.split(',')[0].trim();
  }

  const realIP = headers.get('x-real-ip');
  if (realIP) {
    return realIP;
  }

  const cfConnectingIP = headers.get('cf-connecting-ip'); // Cloudflare
  if (cfConnectingIP) {
    return cfConnectingIP;
  }

  // Fallback (not reliable in production)
  return 'unknown';
}

/**
 * Create rate limit response headers
 */
export function createRateLimitHeaders(result: RateLimitResult): Headers {
  const headers = new Headers();

  headers.set('X-RateLimit-Limit', result.remaining.toString());
  headers.set('X-RateLimit-Remaining', result.remaining.toString());
  headers.set('X-RateLimit-Reset', new Date(result.resetAt).toISOString());

  if (result.retryAfter) {
    headers.set('Retry-After', result.retryAfter.toString());
  }

  return headers;
}

/**
 * Check if request is rate limited
 * Returns true if allowed, false if blocked
 */
export function checkRateLimit(
  rateLimiter: RateLimiter,
  identifier: string
): { allowed: boolean; headers: Headers; result: RateLimitResult } {
  const result = rateLimiter.check(identifier);
  const headers = createRateLimitHeaders(result);

  return {
    allowed: result.allowed,
    headers,
    result,
  };
}
